Allow apps to have their own manifest files

gko opened the pull request:
So that you save a particular part of Nextcloud (i.e. notes)

gko committed:
Allow apps to have their own manifest files

So that you save a particular part of Nextcloud (i.e. notes)

Signed-off-by: Konstantin Gorodinskii <mail@konstantin.io>

kesselb commented on lib/private/Setup.php:
```diff
@@ -523,6 +523,8 @@ public static function updateHtaccess() {
 			$content .= "\n  RewriteCond %{REQUEST_FILENAME} !core/img/favicon.ico$";
 			$content .= "\n  RewriteCond %{REQUEST_FILENAME} !core/img/manifest.json$";
+			$content .= "\n  RewriteCond %{REQUEST_FILENAME} !apps/[\w-_]+/img/manifest.json$";
+			$content .= "\n  RewriteCond %{REQUEST_FILENAME} !custom_apps/[\w-_]+/img/manifest.json$";
```
This one is not fixed. People can pick another path.

kesselb replied:
```
			foreach (\OC::$APPSROOTS as ['url' => $url]) {
				$content .= "\n  RewriteCond %{REQUEST_FILENAME} !" . trim($url, '/') . "/[\w-_]+/img/manifest.json$";
			}
```

This one should support custom app paths.

gko replied:
changed it

kesselb reviewed (commented):
Thanks for your pr :+1: 

Would you mind to add some insights about this change? A `manifest.json` per app is necessary to ...

gko committed:
we need to keep add hash to manifest file if it exists

Signed-off-by: Konstantin Gorodinskii <mail@konstantin.io>

gko commented:
@kesselb Thank you for the quick response. I wrote the reasoning in the second [PR](https://github.com/nextcloud/notes/pull/422)

Basically I want to be able to access a part of Nextcloud directly via «Add to Home screen» feature of a mobile browser.

Right now `manifest` is always rewritten by default app `theming`. It checks whether there is an existing file `apps/<app>/img/manifest.json`. Otherwise it generates its own `manifest` in the [ThemingDefaults.php](https://github.com/nextcloud/server/blob/2442d78696313744bae8170c1b547f3b50ca8fcd/apps/theming/lib/ThemingDefaults.php#L347-L355) file:
```php
if ($image === 'manifest.json') {
	try {
		$appPath = $this->appManager->getAppPath($app);
		if (file_exists($appPath . '/img/manifest.json')) {
			return false;
		}
	} catch (AppPathNotFoundException $e) {}
	$route = $this->urlGenerator->linkToRoute('theming.Theming.getManifest');
}
```

The problem here is that half of the feature was already there (if the manifest exists it stays in the `meta` tag) but it's then blocked (redirected) by `.htaccess` file.

That's why I added those two rules.

The second part is that there is never a hash added to manifest file if it exists. I fixed it in the second commit.

gary-kim commented:
Wrong button :)

gko committed:
more generic solution

Signed-off-by: Konstantin Gorodinskii <mail@konstantin.io>

kesselb commented on apps/theming/lib/ThemingDefaults.php:
```diff
@@ -348,7 +348,7 @@ public function replaceImagePath($app, $image) {
 				$appPath = $this->appManager->getAppPath($app);
 				if (file_exists($appPath . '/img/manifest.json')) {
-					return false;
+					return $this->urlGenerator->linkTo($app, 'img/manifest.json') . '?v=' . $cacheBusterValue;
```
Not sure about this code. It will only run with the theming app enabled. Do we need the cachebuster for manifest.json?

kesselb replied:
Code seem to be correct. Theming app should not replace the manifest.json. I would not change that.

gko replied:
The thing is that in the `layout.user.php` there is no hash on manifest.json.

I'll need to do couple of tests in mobile browsers to check if they fetch everytime the new manifest.

gko replied:
ok, I've tested on chrome (Android) and Safari (iOS). They seem to refetch the manifest file.
So I reverted the commit.

kesselb commented:
Ah. Did some testing. This approach will only work for instances with mod_rewrite enabled. It does not work with nginx. RewriteRule will tell the webserver to not forward the request for /apps/notes/img/manifest.json to nextcloud and serve it statically. That's fine but we probably need a new rule for nginx. Otherwise nextcloud must be able to serve these files (but that includes booting up everything) and serve the file. That's already there for many static assets (e.g. a generic route to serve js or css from any app). 

Let me ping @rullzer and @juliushaertl to ask them how to do it. The use case here is to ship a custom manifest.json for a app. URL generation seems to work already. Is this something that apps should implement theirself? Or should we tell the webservers to just serve any of these manifest.json files.

```
    location ~ manifest\.json$ {
        try_files $uri =404;
    }
```

That worked for me and will not break the theming manifest (because the url is something like `/apps/theming/manifest?v=0`). But I missing the big picture and don't know how this should integrate with theming ;)

gko committed:
Revert "we need to keep add hash to manifest file if it exists"

This reverts commit d1a8af54889206951ed08c75309a21b7ad40332e.

Signed-off-by: Konstantin Gorodinskii <mail@konstantin.io>

gko commented:
@kesselb thank you for testing on `nginx`. I created a [sandbox](https://github.com/gko/nextcloud-sandbox) to test and develop nextcloud apps. But it uses the image with apache by default.

Also would it be possible to backport it to 16.x and 17.x once we manage to make it work everywhere?

kesselb commented on lib/private/Setup.php:
```diff
@@ -523,6 +523,9 @@ public static function updateHtaccess() {
 			$content .= "\n  RewriteCond %{REQUEST_FILENAME} !core/img/favicon.ico$";
 			$content .= "\n  RewriteCond %{REQUEST_FILENAME} !core/img/manifest.json$";
+			foreach (\OC::$APPSROOTS as ['url' => $url]) {
+				$content .= "\n  RewriteCond %{REQUEST_FILENAME} !" . trim($url, '/') . "/[\w-_]+/img/manifest.json$";
```
Do you think we could replace the foreach with a more generic RewriteCond. If someone is adding a new app_path he will probably not regenerate the htaccess. Such issues are hard to spot even if we document it somewhere. I'm not sure yet. Thats a obscure edge case of course :thinking:

rullzer commented:
If we do this we should make it a generic route that is handled in some controller and then depending on what app you pass to it. Doing .htaccess magic is to error prone in my opinion. But lets revisit this when we start 19 development.

kesselb commented:
> If we do this we should make it a generic route that is handled in some controller and then depending on what app you pass to it.

And bootstrapping the whole nextcloud stack for serving a static file. His approach is to exclude `manifest.json` and let the webserver handle it. I'm sure we can replace foreach appsroots with a generic rule for `manifest.json`.

skjnldsv commented:
Bump?

szaimen commented:
@gko any update here? :)

szaimen commented:
Superseded by https://github.com/nextcloud/server/pull/28459
